File: LuaJIT/Bisection.lua
local math = math
local huge = math.huge

return function (f, a, b, eps, maxIter)
    eps = eps or 1e-12
    maxIter = maxIter or 1e8

    local fA = f(a)
    local fB = f(b)

    if (fA > 0) == (fB > 0) then
        return huge
    end

    if fA == 0 then
        return a
    elseif fB == 0 then
        return b
    end

    for i = 1, maxIter, 1 do
        if b - a <= eps then
            break
        end
        local c = (a + b) / 2

        if (f(a) > 0) ~= (f(c) > 0) then
            b = c
        else
            a = c
        end
    end

    return (a + b) / 2
end

File: Python-Numba/bisection.py
import math

from numba import jit

inf = math.inf
abs_f = abs
range_f = range


@jit
def bisection_method(f, a, b, eps=1e-12, max_iter=10**8):
    f_a = f(a)
    f_b = f(b)

    if (f_a > 0) == (f_b > 0):
        return inf

    if f_a == 0:
        return a
    elif f_b == 0:
        return b

    for i in range_f(max_iter):
        if abs_f(b - a) <= eps:
            break

        c = (a + b) / 2

        if (f(a) > 0) != (f(c) > 0):
            b = c
        else:
            a = c

    return (a - b) / 2
